kafka/channel/pkg/dispatcher/dispatcher.go

```diff
@@ -138,41 +136,61 @@ func (c consumerMessageHandler) Handle(ctx context.Context, consumerMessage *sar
 	if message.ReadEncoding() == binding.EncodingUnknown {
 		return false, errors.New("received a message with unknown encoding")
 	}
-	var destination *url.URL
-	if !c.sub.SubscriberURI.IsEmpty() {
-		destination = c.sub.SubscriberURI.URL()
-	}
-	var reply *url.URL
-	if !c.sub.ReplyURI.IsEmpty() {
-		reply = c.sub.ReplyURI.URL()
-	}
-	var deadLetter *url.URL
-	if c.sub.Delivery != nil && c.sub.Delivery.DeadLetterSink != nil && !c.sub.Delivery.DeadLetterSink.URI.IsEmpty() {
-		deadLetter = c.sub.Delivery.DeadLetterSink.URI.URL()
-	}
+
 	c.logger.Debug("Going to dispatch the message",
 		zap.String("topic", consumerMessage.Topic),
-		zap.String("sub", string(c.sub.UID)),
+		zap.Any("destination", c.sub.SubscriberURI),
+		zap.Any("reply", c.sub.ReplyURI),
+		zap.Any("delivery", c.sub.Delivery),
 	)
 
 	ctx, span := startTraceFromMessage(c.logger, ctx, message, consumerMessage.Topic)
 	defer span.End()
 
-	err := c.dispatcher.DispatchMessage(ctx, message, nil, destination, reply, deadLetter)
+	var DLS *url.URL
+	if c.sub.Delivery != nil && c.sub.Delivery.DeadLetterSink != nil && c.sub.Delivery.DeadLetterSink.URI != nil {
+		DLS = (*url.URL)(c.sub.Delivery.DeadLetterSink.URI)
+	}
+
+	err := c.dispatcher.DispatchMessage(ctx,
+		message,
+		nil,
+		(*url.URL)(c.sub.SubscriberURI),
+		(*url.URL)(c.sub.ReplyURI),
+		DLS,
+	)
+
 	// NOTE: only return `true` here if DispatchMessage actually delivered the message.
 	return err == nil, err
 }
 
 var _ kafka.KafkaConsumerHandler = (*consumerMessageHandler)(nil)
 
-type subscription struct {
+type Config struct {
+	// The configuration of each channel in this handler.
+	ChannelConfigs []ChannelConfig
+}
+
+type ChannelConfig struct {
+	Namespace     string
+	Name          string
+	HostName      string
+	Subscriptions []Subscription
+}
+
+type Subscription struct {
 	eventingduck.SubscriberSpec
```

Comment:
Why that? use this fanout config https://github.com/knative/eventing/blob/master/pkg/channel/fanout/fanout_message_handler.go#L45 and then this function https://github.com/knative/eventing/blob/master/pkg/channel/fanout/fanout_message_handler.go#L93 does the job for you (so you can remove all the hacky code above for conversions)

Reply:
I opened an issue for this: https://github.com/knative/eventing-contrib/issues/1413